File: test_sha1.py
import hashlib
import io
import zipfile

def compute_sha1_stream(data):
    digest = hashlib.sha1()
    digest.update(data)
    return digest.hexdigest()

def read_limited_bytes(data, limit):
    buffer = bytearray()
    total = 0
    stream = io.BytesIO(data)
    
    while True:
        chunk = stream.read(8192)
        if not chunk:
            break
        total += len(chunk)
        if total > limit:
            break
        buffer.extend(chunk)
    return bytes(buffer)

data = b"A" * (25 * 1024 * 1024 + 10) # 25MB + 10 bytes

sha1_stream = compute_sha1_stream(data)
sha1_limited = compute_sha1_stream(read_limited_bytes(data, 25 * 1024 * 1024))

print("Stream:", sha1_stream)
print("Limited:", sha1_limited)

File: test_sha1.kt
import java.security.MessageDigest
import java.io.ByteArrayInputStream
import java.io.InputStream

fun computeSha1(bytes: ByteArray): String {
    val digest = MessageDigest.getInstance("SHA-1")
    val hash = digest.digest(bytes)
    return hash.joinToString("") { "%02x".format(it) }
}

fun computeSha1Stream(inputStream: InputStream): String {
    val digest = MessageDigest.getInstance("SHA-1")
    val buffer = ByteArray(8192)
    var bytesRead: Int
    while (inputStream.read(buffer).also { bytesRead = it } != -1) {
        digest.update(buffer, 0, bytesRead)
    }
    val hash = digest.digest()
    return hash.joinToString("") { "%02x".format(it) }
}

fun main() {
    val data = "Hello World".toByteArray()
    println(computeSha1(data))
    println(computeSha1Stream(ByteArrayInputStream(data)))
}
